pypeit/wavecalib.py

```diff
@@ -6,44 +6,236 @@
+                _d.append({key: self[key]})
+            else: # Add to header of the spat_id image
+                _d[0][key] = self[key]
+        # Return
+        return _d
+
+    @classmethod
+    def _parse(cls, hdu, ext=None, transpose_table_arrays=False, debug=False,
+               hdu_prefix=None):
+        # Grab everything but the bspline's
+        _d, dm_version_passed, dm_type_passed = super(WaveCalib, cls)._parse(hdu)
+        # Now the wave_fits
+        list_of_wave_fits = []
+        spat_ids = []
+        for ihdu in hdu:
+            if 'WAVEFIT' in ihdu.name:
+                # Allow for empty
+                if len(ihdu.data) == 0:
+                    iwavefit = wv_fitting.WaveFit()
+                else:
+                    iwavefit = wv_fitting.WaveFit.from_hdu(ihdu)
+                    if iwavefit.version != wv_fitting.WaveFit.version:
+                        msgs.warn("Your WaveFit is out of date!!")
+                    # Grab PypeItFit (if it exists)
+                    hdname = ihdu.name.replace('WAVEFIT', 'PYPEITFIT')
+                    if hdname in [khdu.name for khdu in hdu]:
+                        iwavefit.pypeitfit = fitting.PypeItFit.from_hdu(hdu[hdname])
+                list_of_wave_fits.append(iwavefit)
+                # Grab SPAT_ID for checking
+                i0 = ihdu.name.find('ID-')
+                i1 = ihdu.name.find('_WAV')
+                spat_ids.append(int(ihdu.name[i0+3:i1]))
+            elif ihdu.name == 'PYPEITFIT': # 2D fit
+                _d['wv_fit2d'] = fitting.PypeItFit.from_hdu(ihdu)
+        # Check
+        if spat_ids != _d['spat_id'].tolist():
+            msgs.error("Bad parsing of the MasterFlat")
+        # Finish
+        _d['wv_fits'] = np.asarray(list_of_wave_fits)
+        return _d, dm_version_passed, dm_type_passed
+
+    @property
+    def par(self):
+        return json.loads(self.strpar)
+
+    def chk_synced(self, slits):
+        """
+        Confirm the slits in WaveCalib are aligned to that in SlitTraceSet
+
+        Barfs if not
+
+        Args:
+            slits (:class:`pypeit.slittrace.SlitTraceSet`):
+
+        """
+        if not np.array_equal(self.spat_id, slits.spat_id):
+            msgs.error("Your wvcalib solutions are out of sync with your slits.  Remove Masters and start from scratch")
+
+    def build_waveimg(self, tilts, slits, spat_flexure=None):
+        """
+        Main algorithm to build the wavelength image
+
+        Only applied to good slits, which means any non-flagged or flagged
+         in the exclude_for_reducing list
+
+        Args:
+            tilts (`numpy.ndarray`_):
+                Image holding tilts
+            slits (:class:`pypeit.slittrace.SlitTraceSet`):
+            spat_flexure (float, optional):
+
+        Returns:
+            `numpy.ndarray`_: The wavelength image.
+        """
+        # Setup
+        #ok_slits = slits.mask == 0
+        bpm = slits.mask.astype(bool)
```

Comment:
I really like the changes made throughout the code that make it clear is mask is a `gpm` or `bpm` - I wonder if all DataContainers (including slits, in this case) should all be updated to be explicit about the type of mask? I think this will help us to avoid making mistakes in the code.

Reply:
definitely.  but not in this PR